compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs

```diff
@@ -411,6 +411,31 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     ty::PredicateKind::Clause(ty::ClauseKind::Trait(trait_predicate)) => {
                         let trait_predicate = bound_predicate.rebind(trait_predicate);
                         let trait_predicate = self.resolve_vars_if_possible(trait_predicate);
+                        let trait_ref = trait_predicate.to_poly_trait_ref();
+
+                        if let ty::Closure(closure_def_id, closure_args) = *trait_ref.self_ty().skip_binder().kind()
+                            && let Some(expected_kind) = self.tcx.fn_trait_kind_from_def_id(trait_ref.def_id())
+                            && let Some(found_kind) = self.closure_kind(closure_args)
+                            && !found_kind.extends(expected_kind)
+                            && let sig = closure_args.as_closure().sig()
+                            && self.can_sub(
+                                obligation.param_env,
+                                trait_ref,
+                                sig.map_bound(|sig| {
+                                    ty::TraitRef::new(
+                                        tcx,
+                                        trait_ref.def_id(),
+                                        [trait_ref.self_ty().skip_binder(), sig.inputs()[0]],
+                                    )
+                                }),
+                            )
+                        {
+                            let mut err = self.report_closure_error(&obligation, closure_def_id, found_kind, expected_kind);
+                            self.note_obligation_cause(&mut err, &obligation);
+                            self.point_at_returns_when_relevant(&mut err, &obligation);
+                            err.emit();
+                            return;
+                        }
```

Comment:
can you put this into a separate funciton, something like `if self.emit_specialized_closure_kind_error() { return }`?